src/linear_algebra/pi_tensor_product.lean

```diff
@@ -354,9 +357,87 @@ theorem lift.unique {φ' : (⨂[R] i, s i) →ₗ[R] E} (H : ∀ f, φ' (tprod R
   φ' = lift φ :=
 lift.unique' (multilinear_map.ext H)
 
+@[simp]
+theorem lift_symm (φ' : (⨂[R] i, s i) →ₗ[R] E) : lift.symm φ' = φ'.comp_multilinear_map (tprod R) :=
+rfl
+
+@[simp]
 theorem lift_tprod : lift (tprod R : multilinear_map R s _) = linear_map.id :=
 eq.symm $ lift.unique' rfl
 
+section
+variables (R M)
+/-- Re-index the components of the tensor power by `e`.
+
+For simplicity, this is defined only for homogeneously- (rather than dependently-) typed components.
+-/
+def reindex (e : ι ≃ ι₂) : ⨂[R] i : ι, M ≃ₗ[R] ⨂[R] i : ι₂, M :=
+linear_equiv.of_linear
+  ((lift.symm.trans $
+    multilinear_map.dom_dom_congr_linear_equiv M (⨂[R] i : ι₂, M) R R e.symm).trans
+      lift (linear_map.id))
+  ((lift.symm.trans $
+    multilinear_map.dom_dom_congr_linear_equiv M (⨂[R] i : ι, M) R R e).trans
+      lift (linear_map.id))
+  (by { ext, simp })
+  (by { ext, simp })
+end
+
+@[simp] lemma reindex_tprod (e : ι ≃ ι₂) (f : Π i, M) :
+  reindex R M e (tprod R f) = tprod R (λ i, f (e.symm i)) :=
+lift.tprod f
+
+@[simp] lemma reindex_comp_tprod (e : ι ≃ ι₂) :
+  (reindex R M e : ⨂[R] i : ι, M →ₗ[R] ⨂[R] i : ι₂, M).comp_multilinear_map (tprod R) =
+    (tprod R : multilinear_map R (λ i, M) _).dom_dom_congr e.symm :=
+multilinear_map.ext $ reindex_tprod e
+
+@[simp] lemma lift_comp_reindex (e : ι ≃ ι₂) (φ : multilinear_map R (λ _ : ι₂, M) E) :
+  (lift φ).comp ↑(reindex R M e) = lift (φ.dom_dom_congr e.symm) :=
+by { ext, simp, }
+
+@[simp]
+lemma lift_reindex (e : ι ≃ ι₂) (φ : multilinear_map R (λ _, M) E) (x : ⨂[R] i, M) :
+  lift φ (reindex R M e x) = lift (φ.dom_dom_congr e.symm) x :=
+linear_map.congr_fun (lift_comp_reindex e φ) x
+
+@[simp] lemma reindex_trans (e : ι ≃ ι₂) (e' : ι₂ ≃ ι₃) :
+  (reindex R M e).trans (reindex R M e') = reindex R M (e.trans e') :=
+linear_equiv.injective_to_linear_map begin
+  ext f,
+  simp only [linear_equiv.trans_apply, linear_equiv.coe_coe, reindex_tprod,
+    linear_map.coe_comp_multilinear_map, function.comp_app, multilinear_map.dom_dom_congr_apply,
+    reindex_comp_tprod],
+  congr,
+end
```

Comment:
Let's move `begin` to its own line to make clearer that this is a tactic block. (Alternatively, you can put `apply` inside the tactic block, right?)
```suggestion
linear_equiv.injective_to_linear_map
  begin
    ext f,
    simp only [linear_equiv.trans_apply, linear_equiv.coe_coe, reindex_tprod,
      linear_map.coe_comp_multilinear_map, function.comp_app, multilinear_map.dom_dom_congr_apply,
      reindex_comp_tprod],
    congr,
  end
```

Reply:
Would you prefer `apply` in the tactic block?

Reply:
```suggestion
begin
  apply linear_equiv.injective_to_linear_map,
  ext f,
  simp only [linear_equiv.trans_apply, linear_equiv.coe_coe, reindex_tprod,
    linear_map.coe_comp_multilinear_map, function.comp_app, multilinear_map.dom_dom_congr_apply,
    reindex_comp_tprod],
  congr,
end
```